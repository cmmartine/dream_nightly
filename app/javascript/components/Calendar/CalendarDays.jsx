import React from "react";

export default function CalendarDays(props) {
  const {
    calendarDay,
    daysInfo,
    handleDateChange
  } = props;

  const renderDaysInMonth = () => {
    if (!daysInfo) return null;

    const firstDayOfWeek = daysInfo[0].day_of_week;

    const emptyCells = Array.from({ length: firstDayOfWeek }, (_, index) => (
      <div key={`empty-${index}`} className="calendar-day empty"></div>
    ));

    const dayCells = daysInfo.map((day) => (
      <div className={`calendar-day-container`}>
        <button
          key={day.num}
          className={`calendar-day ${day.has_dreams ? 'highlight' : ''} ${day.is_today ? 'calendar-today' : ''} ${day.num === calendarDay ? 'highlight-selected' : ''}`}
          aria-label='Change the day'
          onClick={() => {
            handleDateChange({ newDay: day.num })
          }}
        >
          {day.num}
        </button>
      </div>
    ));

    return [...emptyCells, ...dayCells];
  };

  return (
    <div className='calendar-days-view-container'>
      <div className='calendar-weekdays-container'>
        <div className='calendar-weekday'>Sun</div>
        <div className='calendar-weekday'>Mon</div>
        <div className='calendar-weekday'>Tue</div>
        <div className='calendar-weekday'>Wed</div>
        <div className='calendar-weekday'>Thu</div>
        <div className='calendar-weekday'>Fri</div>
        <div className='calendar-weekday'>Sat</div>
      </div>
      <div className="calendar-days-container">{renderDaysInMonth()}</div>
    </div>
  );
}