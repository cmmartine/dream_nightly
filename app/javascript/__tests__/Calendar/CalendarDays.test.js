import { render, screen, fireEvent } from "@testing-library/react";
import CalendarDays from "../../components/Calendar/CalendarDays";

describe('CalendarDays', () => {

  const calendarDay = 2;
  const handleDateChange= jest.fn();

  const daysInfo = [
    {
      num: 1,
      has_dreams: true,
      day_of_week: 3,
      is_today: true
    },
    {
      num: 2,
      has_dreams: false,
      day_of_week: 4,
      is_today: false
    }
  ]

  function renderCalendarDays() {
    return render(
      <CalendarDays calendarDay={calendarDay} daysInfo={daysInfo} handleDateChange={handleDateChange}/>
    )
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders a button for each day passed in', () => {
    renderCalendarDays();

    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    expect(screen.getAllByRole('button').length).toBe(2);
  })

  it('renders correctly', () => {
    const calendarDays = renderCalendarDays();
    expect(calendarDays).toMatchSnapshot();
  });

  it('calls handleDateChange when a number button is clicked', () => {
    renderCalendarDays();

    fireEvent.click(screen.getAllByRole('button')[0]);
    expect(handleDateChange).toHaveBeenCalled();
  });
});