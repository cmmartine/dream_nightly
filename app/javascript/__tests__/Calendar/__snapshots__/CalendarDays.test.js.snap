// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarDays renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="calendar-days-view-container"
      >
        <div
          class="calendar-weekdays-container"
        >
          <div
            class="calendar-weekday"
          >
            Sun
          </div>
          <div
            class="calendar-weekday"
          >
            Mon
          </div>
          <div
            class="calendar-weekday"
          >
            Tue
          </div>
          <div
            class="calendar-weekday"
          >
            Wed
          </div>
          <div
            class="calendar-weekday"
          >
            Thu
          </div>
          <div
            class="calendar-weekday"
          >
            Fri
          </div>
          <div
            class="calendar-weekday"
          >
            Sat
          </div>
        </div>
        <div
          class="calendar-days-container"
        >
          <div
            class="calendar-day empty"
          />
          <div
            class="calendar-day empty"
          />
          <div
            class="calendar-day empty"
          />
          <div
            class="calendar-day-container"
          >
            <button
              aria-label="Change the day"
              class="calendar-day highlight calendar-today "
            >
              1
            </button>
          </div>
          <div
            class="calendar-day-container"
          >
            <button
              aria-label="Change the day"
              class="calendar-day   highlight-selected"
            >
              2
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="calendar-days-view-container"
    >
      <div
        class="calendar-weekdays-container"
      >
        <div
          class="calendar-weekday"
        >
          Sun
        </div>
        <div
          class="calendar-weekday"
        >
          Mon
        </div>
        <div
          class="calendar-weekday"
        >
          Tue
        </div>
        <div
          class="calendar-weekday"
        >
          Wed
        </div>
        <div
          class="calendar-weekday"
        >
          Thu
        </div>
        <div
          class="calendar-weekday"
        >
          Fri
        </div>
        <div
          class="calendar-weekday"
        >
          Sat
        </div>
      </div>
      <div
        class="calendar-days-container"
      >
        <div
          class="calendar-day empty"
        />
        <div
          class="calendar-day empty"
        />
        <div
          class="calendar-day empty"
        />
        <div
          class="calendar-day-container"
        >
          <button
            aria-label="Change the day"
            class="calendar-day highlight calendar-today "
          >
            1
          </button>
        </div>
        <div
          class="calendar-day-container"
        >
          <button
            aria-label="Change the day"
            class="calendar-day   highlight-selected"
          >
            2
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
